From cdadc6a0870bc7a6d4cab05a99276f0817feed83 Mon Sep 17 00:00:00 2001
From: Takuro Ashie <ashie@clear-code.com>
Date: Tue, 27 Feb 2018 09:48:38 +0900
Subject: [PATCH 2/2] OmxDataDecoder: Fix a stall issue on shutting down

Because the shutdown closure awaits finishing itself by
TaskQueue::AwaitShutdownAndIdle(), the function blocks infinitely.

The code is wrongly introduced at the following commit:

  * https://bugzilla.mozilla.org/show_bug.cgi?id=1319987
    * https://hg.mozilla.org/mozilla-central/rev/b2171e3e8b69

It introduces async shutdown mechanism so that calling
AwaitShutdownAndIdle() isn't needed anymore.

Signed-off-by: Takuro Ashie <ashie@clear-code.com>
---
 dom/media/platforms/omx/OmxDataDecoder.cpp | 11 ++---------
 dom/media/platforms/omx/OmxDataDecoder.h   |  1 -
 2 files changed, 2 insertions(+), 10 deletions(-)

diff --git a/dom/media/platforms/omx/OmxDataDecoder.cpp b/dom/media/platforms/omx/OmxDataDecoder.cpp
index 4e8230dd1ec0..2e0f860d660b 100644
--- a/dom/media/platforms/omx/OmxDataDecoder.cpp
+++ b/dom/media/platforms/omx/OmxDataDecoder.cpp
@@ -279,24 +279,17 @@ OmxDataDecoder::DoAsyncShutdown()
              self->mWatchManager.Shutdown();
              self->mOmxLayer = nullptr;
              self->mMediaDataHelper = nullptr;
-
              self->mShuttingDown = false;
-             self->mOmxTaskQueue->BeginShutdown();
-             self->mOmxTaskQueue->AwaitShutdownAndIdle();
-             self->mShutdownPromise.Resolve(true, __func__);
            },
            [self] () {
              self->mOmxLayer->Shutdown();
              self->mWatchManager.Shutdown();
              self->mOmxLayer = nullptr;
              self->mMediaDataHelper = nullptr;
-
              self->mShuttingDown = false;
-             self->mOmxTaskQueue->BeginShutdown();
-             self->mOmxTaskQueue->AwaitShutdownAndIdle();
-             self->mShutdownPromise.Resolve(true, __func__);
            });
-  return mShutdownPromise.Ensure(__func__);
+
+  return mOmxTaskQueue->BeginShutdown();
 }
 
 void
diff --git a/dom/media/platforms/omx/OmxDataDecoder.h b/dom/media/platforms/omx/OmxDataDecoder.h
index 9bc480670e07..2d5b36c36da5 100644
--- a/dom/media/platforms/omx/OmxDataDecoder.h
+++ b/dom/media/platforms/omx/OmxDataDecoder.h
@@ -188,7 +188,6 @@ protected:
   MozPromiseHolder<DecodePromise> mDecodePromise;
   MozPromiseHolder<DecodePromise> mDrainPromise;
   MozPromiseHolder<FlushPromise> mFlushPromise;
-  MozPromiseHolder<ShutdownPromise> mShutdownPromise;
   // Where decoded samples will be stored until the decode promise is resolved.
   DecodedData mDecodedData;
 
-- 
2.14.1

